OSSM-4774 Use internal registry of OCP

fjglira opened the pull request:
The goal is to avoid the use of quay as a registry to build and push images to be tested for the operator.

Changes done:
* Rebuild the entire testing common script to separate all the steps into functions to easily understand the goal of each step and also to separate the actual test from the setup steps
* Add verification to get the internal registry
* Login to the internal registry
* Run docker build and push using the HUB variable to override the value in the make file

mattmahoneyrh commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,101 +14,146 @@
+log_internal_registry
+build_and_push_image
+deploy_operator
+
```
For operator test reuse with downstream pipeline, consider providing the ability to only run the tests. The operator deploy via subscription could be done within this suite or done so prior to running the test.

fjglira replied:
Ok, that sounds reasonable to me. I'll add the capability to skip setup

fjglira commented:
I'm making some changes to avoid to kind execution to fail because of the use of the internal registry (It's not needed for kind this steps)

fjglira commented:
/test ci/prow/integration-kind

fjglira commented:
/test integration-kind

fjglira commented:
/test integration-ocp

fjglira commented:
/test integration-ocp

fjglira commented:
/test integration-ocp

fjglira commented:
/test integration-ocp

fjglira commented:
/test integration-ocp

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
-# To be able to run this script it's needed to pass the flag --ocp or --kind
-set -eu -o pipefail
+# To be able to run this script, it's needed to pass the flag --ocp or --kind
+set -eux -o pipefail
```
You can remove the `-x` in the final version.

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
+  echo "Using command: ${COMMAND}"
 
-BRANCH="${BRANCH:-maistra-3.0}"
+  BRANCH="${BRANCH:-maistra-3.0}"
```
Is this variable used anywhere?

fjglira replied:
The only place where I see this variable used is in the target bundle-publish, which we do not use for testing. Deleting

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
-    local POD_NAME=$2
-    local DEPLOYMENT_NAME=$3
+  # Set a default value for a variable
+  : "${RUN_ONLY_TEST:="false"}"
```
If you change the assignment, perhaps it's more clear to the reader, making the comment above not necessary.
e.g. `RUN_ONLY_TEST=${RUN_ONLY_TEST:-"false"}

Although in my opinion this variable should be called `SKIP_BUILD` and should be passed via command line arguments (e.g. `--skip-build`).

fjglira replied:
Added

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
+get_internal_registry() {
+  # Validate that the internal registry is running, configure the variable to be used in the Makefile. 
+  # If there is no internal registry, we will use the public one defined in Makefile by default
+
```
Why? I thought the idea was to rely completely on the presence of the internal registry. Any reason to not to?

fjglira replied:
because if not exist we will need to install it, in theory this case will not happen because our ci is creating the ocp cluster with the internal registry by default

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
+  # If there is no internal registry, we will use the public one defined in Makefile by default
+
+  # Check if the registry pods are running
+  $COMMAND get pods -n openshift-image-registry --no-headers | grep -v "Running" && exit 1
```
Here we are doing it differently than what's commented above. If the registry pod does not exist, the test exits with an error. That's fine, although I think we should emit a message. Something like: "It looks like the OCP image registry is not deployed. This tests requires it. Aborting."

fjglira replied:
Ummm ok ok, understood

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
-    echo "Check Deployment Available: NAMESPACE: \"${NS}\"   DEPLOYMENT NAME: \"${DEPLOYMENT_NAME}\""
-    $COMMAND  wait deployment "${DEPLOYMENT_NAME}" -n "${NS}" --for condition=Available=True --timeout=${TIMEOUT}
+  # Patch to activate the default route to the registry
+  $COMMAND patch configs.imageregistry.operator.openshift.io/cluster --patch '{"spec":{"defaultRoute":true}}' --type=merge
```
Instead of blindly patching it, should we query if the default route already exists? If it does we skip the patch and the sleep.

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
+  oc get routes -n openshift-image-registry
+  URL=$($COMMAND get route default-route -n openshift-image-registry --template='{{ .spec.host }}')
+  # Hub will be equal to the route url/project-name(NameSpace) 
+  export HUB="${URL}/$NAMESPACE"
```
nit: I'm in favor of using `${}` everywhere to reference variables.

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
+
+  # Create namespace where operator will be located
+  # This is needed because the roles are created in the namespace where the operator is deployed
+  oc create namespace "$NAMESPACE" 
```
I know that at this point, `oc` is going to be used. But I think we should keep using `${COMMAND}` everywhere. Plus, maybe we should append ` || true` so that the command doesn't fail if the namespace already exists.

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
+  echo "Image base: ${IMAGE_BASE}"
+  echo " Tag: ${TAG}"
+  IMAGE=${HUB}/${IMAGE_BASE}:${TAG} make docker-build docker-push
+  exit 0
```
leftover?

fjglira replied:
Yes :)

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,87 +14,141 @@
+  $COMMAND patch configs.imageregistry.operator.openshift.io/cluster --patch '{"spec":{"defaultRoute":true}}' --type=merge
+
+  # Get the registry route
+  sleep 30 # TODO: improve waiting time to avoid hardcoded sleep
```
You can borrow the `until` logic that's used below.

jwendell reviewed (commented):
It's looking good! I made a few comments though.

fjglira commented:
@jwendell PR updated please take a look

luksa commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,118 +14,216 @@
+    ${COMMAND} get ns "${CONTROL_PLANE_NS}" >/dev/null 2>&1 || ${COMMAND} create namespace "${CONTROL_PLANE_NS}"
+    sed -e "s/version:.*/version: ${ver}/g" "${ISTIO_MANIFEST}" | ${COMMAND} apply -f - -n "${CONTROL_PLANE_NS}"
+
+    # older versions of istio don't run on OpenShift without anyuid and privileged SCC
+    if [ "${OCP}" == "true" ] && [ "$ver" != "latest" ]; then
+        echo "Configure anyuid and privileged SCC in ${CONTROL_PLANE_NS}"
+        ${COMMAND}  adm policy add-scc-to-group anyuid "system:serviceaccounts:${CONTROL_PLANE_NS}"
+        ${COMMAND}  adm policy add-scc-to-group privileged "system:serviceaccounts:${CONTROL_PLANE_NS}"
+    fi
```
This is no longer needed.

luksa commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,118 +14,216 @@
       exit 1
     fi
 
+    if [ "${OCP}" == "true" ] && [ "$ver" != "latest" ]; then
+        echo "Remove anyuid and privileged SCC in ${CONTROL_PLANE_NS}"
+        ${COMMAND}  adm policy remove-scc-from-group anyuid "system:serviceaccounts:${CONTROL_PLANE_NS}"
+        ${COMMAND}  adm policy remove-scc-from-group privileged "system:serviceaccounts:${CONTROL_PLANE_NS}"
+    fi
```
No longer needed.

fjglira replied:
Ok, deleting

fjglira committed:
OSSM-4774 Use internal registry of OCP for integration test

Adding role-binding file for integration test on ocp

fix lint

Improve docker login in integration common test

luksa commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,118 +14,204 @@
+  # Take into count that you will need to add before the registry URL as Insecure registry in "/etc/docker/daemon.json"
+  if [[ ${URL} == *".apps-crc.testing"* ]]; then
+    echo "Executing Docker login to the internal registry"
+    docker login -u $(${COMMAND} whoami) -p $(${COMMAND} whoami -t) ${URL}
```
```suggestion
    ${COMMAND} whoami -t | docker login -u "$(${COMMAND} whoami)" --password-stdin ${URL}
```

luksa replied:
Also, since this entire function is only used with OpenShift, does it really make sense to use `${COMMAND}` instead of simply `oc`? Why make the code harder to read?

fjglira replied:
Was only intended to have everything with the same format. But we can change it to make easy to read

fjglira replied:
Change done and added a message for when the docker login login fails to aware the user that is needed to be added insecure registry when you are running the test locally and not from the prow job

fjglira commented:
Test `ci/prow/integration-ocp` is expected to fail, needs to be merged before this PR: https://github.com/openshift/release/pull/44843 to be able to properly test this PR

fjglira commented:
```
docker push default-route-openshift-image-registry.apps.5cf21f86c742b4ed71e0.hypershift.aws-2.ci.openshift.org/istio-operator/istio-operator:PR-1356
The push refers to repository [default-route-openshift-image-registry.apps.5cf21f86c742b4ed71e0.hypershift.aws-2.ci.openshift.org/istio-operator/istio-operator]
Get "https://default-route-openshift-image-registry.apps.5cf21f86c742b4ed71e0.hypershift.aws-2.ci.openshift.org/v2/": tls: failed to verify certificate: x509: certificate signed by unknown authority
```
When tries to do the push of the image fails because it's not added as insecure registry the default registry

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,118 +14,208 @@
+    # To avoid errors of certificates meanwhile we are pulling the operator image from the internal registry
+    # We need to set image $HUB to a fixed known value
+    # This value always will be equal to the svc url of the internal registry
+    HUB="image-registry.openshift-image-registry.svc:5000/istio-operator"
```
Is this a leftover? It's not clear to me why we need this.

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,118 +14,208 @@
+
+  # Check if default route already exist
+  if [ -z "$(${COMMAND} get route default-route -n openshift-image-registry -o name)" ]; then
+    echo "Route default-route does not exist, so you can perform patching here."
```
This wording is confusing. Are you telling the user they can perform a patch here?

fjglira replied:
My bad, changing the phrase

jwendell commented on tests/integration/common-operator-integ-suite.sh:
```diff
@@ -14,118 +14,208 @@
+  local DEPLOYMENT_NAME=$3
 
-# Main
+  echo "Check POD: NAME SPACE: \"${NS}\"   POD NAME: \"${POD_NAME}\""
```
nit: NAMESPACE (altogether)

fjglira committed:
fix echo log on integration test

fjglira committed:
fix echo log on integration test

fjglira commented:
/retest

jwendell commented:
/retest
